src/liballoc/arc.rs

```diff
@@ -1028,13 +1028,13 @@ impl<T> Weak<T> {
     #[stable(feature = "downgraded_weak", since = "1.10.0")]
     pub fn new() -> Weak<T> {
         unsafe {
-            Weak {
-                ptr: Box::into_raw_non_null(box ArcInner {
-                    strong: atomic::AtomicUsize::new(0),
-                    weak: atomic::AtomicUsize::new(1),
-                    data: uninitialized(),
-                }),
-            }
+            let layout = Layout::new::<ArcInner<T>>();
+            let mut ptr = Global.alloc(layout)
+                .unwrap_or_else(|_| handle_alloc_error(layout))
+                .cast::<ArcInner<T>>();
+            ptr.as_mut().strong = atomic::AtomicUsize::new(0);
+            ptr.as_mut().weak = atomic::AtomicUsize::new(1);
```

Comment:
Using `=` rather than `ptr::write` would be incorrect in a generic context, but I believe it is fine here since `AtomicUsize: !Drop`.

Reply:
Fair point. Replace with `ptr::write`.